functorch/__init__.py

```diff
@@ -91,7 +91,15 @@ def _functorch_str(tensor):
         return _old_str(tensor)
 
     value = _C.get_unwrapped(tensor)
+    # Disable temporarily kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    dl_enabled = _C.tls_set_is_included()
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(False)
     value_repr = repr(value)
+    # Reenable kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(True)
```

Comment:
We probably want to use some "RAII" here: if the `repr` call fails then we want to make sure the state is still OK:
```
try: 
  if (dl_enabled):
        _C._set_dynamic_layer_keys_included(False)
finally:
   if (dl_enabled):
        _C._set_dynamic_layer_keys_included(True)
```

Reply:
I see your point, but just wonder if we really need to do that. If `value_repr = repr(value)` fails then script execution stops. If user does try-except some part of his/her code where `print(t)` errors, catches the error and continue the script, we should ensure that the state is OK and your pattern would totally make sense. I wonder how "typical" this pattern in python development ?

Reply:
> if user does try-except some part of his/her code where print(t) errors, catches the error and continue the script, we should ensure that the state is OK and your pattern would totally make sense.  I wonder how "typical" this pattern in python development ?

You're right that this pattern would probably not be common in user code. There's two concrete use cases for this that I've seen:
1. Someone has a long-running training loop. If an error occurs, I've seen some people do a try-catch and then save their model state in the finally() step so that no training progress is lost.
2. Testing. If someone (really, just us :P) has a test suite that tests tensor printing, if the printing raises an error in a unit test then it'll affect the state of all other subsequent unit tests